peers/app_request_network.go

```diff
@@ -126,95 +135,171 @@ func NewNetwork(
 		return nil, err
 	}
 
-	validatorClient := validators.NewCanonicalValidatorClient(logger, cfg.GetPChainAPI())
+	for _, peer := range manuallyTrackedPeers {
+		logger.Info(
+			"Manually Tracking peer (startup)",
+			zap.Stringer("ID", peer.ID),
+			zap.Stringer("IP", peer.PublicIP),
+		)
+		testNetwork.ManuallyTrack(peer.ID, peer.PublicIP)
+	}
 
-	arNetwork := &appRequestNetwork{
-		network:              testNetwork,
-		handler:              handler,
-		infoAPI:              infoAPI,
-		logger:               logger,
-		lock:                 new(sync.Mutex),
-		validatorClient:      validatorClient,
-		metrics:              metrics,
-		manuallyTrackedPeers: manuallyTrackedPeers,
+	// Connect to a sample of the primary network validators, with connection
+	// info pulled from the info API
+	peers, err := infoAPI.Peers(context.Background(), nil)
+	if err != nil {
+		logger.Error(
+			"Failed to get peers",
+			zap.Error(err),
+		)
+		return nil, err
+	}
+	peersMap := make(map[ids.NodeID]info.Peer)
+	for _, peer := range peers {
+		peersMap[peer.ID] = peer
 	}
+
+	pClient := platformvm.NewClient(cfg.GetPChainAPI().BaseURL)
+	vdrs, err := pClient.GetCurrentValidators(context.Background(), constants.PrimaryNetworkID, nil)
+	if err != nil {
+		logger.Error("Failed to get current validators", zap.Error(err))
+		return nil, err
+	}
+
+	// Sample until we've connected to the target number of bootstrap nodes
+	s := sampler.NewUniform()
+	s.Initialize(uint64(len(vdrs)))
+	numConnected := 0
+	for numConnected < NumBootstrapNodes {
+		if i, ok := s.Next(); ok {
+			if peer, ok := peersMap[vdrs[i].NodeID]; ok {
+				logger.Info(
+					"Manually tracking bootstrap node",
+					zap.Stringer("ID", peer.ID),
+					zap.Stringer("IP", peer.PublicIP),
+				)
+				testNetwork.ManuallyTrack(peer.ID, peer.PublicIP)
+				numConnected++
+			}
+		} else {
+			// If we've sampled all the nodes and still haven't connected to the target number of bootstrap nodes,
+			// then warn and stop sampling by either returning an error or breaking
+			logger.Warn(
+				"Failed to connect to enough bootstrap nodes",
+				zap.Int("targetBootstrapNodes", NumBootstrapNodes),
+				zap.Int("numAvailablePeers", len(peers)),
+				zap.Int("connectedBootstrapNodes", numConnected),
+			)
+			if numConnected == 0 {
+				return nil, fmt.Errorf("failed to connect to any bootstrap nodes")
+			}
+			break
+		}
```

Comment:
This is pretty nitpicky, but I think the bug you had here not breaking out of the loop probably would have been caught faster if you followed the pattern of checking for error cases first, and then continuing function execution as normal, ex: 

```suggestion
		i, ok := s.Next() 
		if !ok {
			// If we've sampled all the nodes and still haven't connected to the target number of bootstrap nodes,
			// then warn and stop sampling by either returning an error or breaking
			logger.Warn(
				"Failed to connect to enough bootstrap nodes",
				zap.Int("targetBootstrapNodes", NumBootstrapNodes),
				zap.Int("numAvailablePeers", len(peers)),
				zap.Int("connectedBootstrapNodes", numConnected),
			)
			if numConnected == 0 {
				return nil, fmt.Errorf("failed to connect to any bootstrap nodes")
			}
			break
		}
	        if peer, ok := peersMap[vdrs[i].NodeID]; ok {
		        logger.Info(
			        "Manually tracking bootstrap node",
			        zap.Stringer("ID", peer.ID),
			        zap.Stringer("IP", peer.PublicIP),
		        )
		        testNetwork.ManuallyTrack(peer.ID, peer.PublicIP)
		        numConnected++
	        }
```

Reply:
I like this better - updated now